test-data/unit/check-python310.test

```diff
@@ -1725,3 +1725,60 @@ def my_func(pairs: Iterable[tuple[S, S]]) -> None:
                 reveal_type(pair)  # N: Revealed type is "Tuple[builtins.int, builtins.int]" \
                                    # N: Revealed type is "Tuple[builtins.str, builtins.str]"
 [builtins fixtures/tuple.pyi]
+
+[case testPartiallyDefinedMatch]
+# flags: --enable-error-code partially-defined
+def f1(a: object) -> None:
+    match a:
+        case [y]: pass
+        case _:
+            y = 1
+            x = 2
+    z = y
+    z = x  # E: Name "x" may be undefined
+
+def f2(a: object) -> None:
+    match a:
+        case [[y] as x]: pass
+        case {"k1": 1, "k2": x, "k3": y}: pass
+        case [0, *x]:
+            y = 2
+        case _:
+            y = 1
+            x = [2]
+    z = x
+    z = y
+
+def f3(a: object) -> None:
+    y = 1
+    match a:
+        case [x]:
+            y = 2
+        # Note the missing `case _:`
+    z = x  # E: Name "x" may be undefined
+    z = y
+
+def f4(a: object) -> None:
+    y = 1
+    match a:
+        case [x]:
+            y = 2
+        case _:
+            assert False, "unsupported"
+    z = x
+    z = y
+
+def f5(a: object) -> None:
+    match a:
+        case tuple(x): pass
+        case _:
+            return
+    y = x
+
+def f6(a: object) -> None:
+    if int():
+        y = 1
+    match a:
+        case _ if y is not None:  # E: Name "y" may be undefined
+            pass
+[builtins fixtures/tuple.pyi]
```

Comment:
This doesn't generate an error, though I think it should:
```py
def f(x: int | str) -> int:
    match x:
        case int():
            y = 1
    return y  # no error
```

Reply:
The exact example you provided generates the error for me. Can you double check how you ran this?

I've added this exact example into the test.

Reply:
Sorry, I was running the wrong code. The example works just fine. However, this related case seems to not work as expected:
```py
def f(x: int | str) -> int:
    match x:
        case int():
            y = 1
        case str():
            y = 2
    return y  # error: Name "y" may be undefined
```

Since the match statement covers all the union items, there shouldn't be an error. Similarly, this generates a false positive:
```py
def f2(x: int | str) -> int:
    if isinstance(x, int):
        y = 1
    elif isinstance(x, str):
        y = 2
    return y  # error: Name "y" may be undefined
```

However, mypy doesn't complain about a missing return statement in these related examples, so mypy can already detect if all union items are covered:
```py
def f3(x: int | str) -> int:
    if isinstance(x, int):
        return 1
    elif isinstance(x, str):
        return 2

def f4(x: int | str) -> int:
    match x:
        case int():
            return 1
        case str():
            return 2
```

Since this also affects `isinstance` checks, this may not be directly related to this PR, and fixed in a separate PR.

Reply:
Good find! 

Agreed, separate PR is probably a better place to do this. I created #13926.